pkg/reconciler/buildrun/resources/service_accounts.go

```diff
@@ -38,70 +39,112 @@ func IsGeneratedServiceAccountUsed(buildRun *buildv1alpha1.BuildRun) bool {
 	return buildRun.Spec.ServiceAccount != nil && buildRun.Spec.ServiceAccount.Generate
 }
 
+// GenerateSA generates a new service account on the fly
+func GenerateSA(ctx context.Context, client client.Client, build *buildv1alpha1.Build, buildRun *buildv1alpha1.BuildRun) (serviceAccount *corev1.ServiceAccount, err error) {
+	serviceAccount = &corev1.ServiceAccount{}
+	err = client.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      GetGeneratedServiceAccountName(buildRun),
+			Namespace: buildRun.Namespace},
+		serviceAccount)
+
+	switch {
+	case err == nil: // if the service account already exists, do nothing and just return it
+		ctxlog.Info(ctx, "serviceAccount for BuildRun already exists", namespace, buildRun.Namespace, name, serviceAccount.Name, "BuildRun", buildRun.Name)
+		return serviceAccount, nil
+
+	case apierrors.IsNotFound(err):
+		// populate the object with well known required fields
+		serviceAccount = &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      GetGeneratedServiceAccountName(buildRun),
+				Namespace: buildRun.Namespace,
+				Labels:    map[string]string{buildv1alpha1.LabelBuildRun: buildRun.Name},
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(buildRun, buildv1alpha1.SchemeGroupVersion.WithKind("BuildRun")),
+				},
+			},
+			AutomountServiceAccountToken: pointer.BoolPtr(false),
+		}
+		ctxlog.Debug(ctx, "automatic generation of service account", namespace, serviceAccount.Namespace, name, serviceAccount.Name)
+
+		// if we didnt have the sa, then create one and ensure the Build secrets are referenced
+		if err := client.Create(ctx, serviceAccount); err != nil {
+			return nil, err
+		}
+		ctxlog.Info(ctx, "created serviceAccount for BuildRun", namespace, buildRun.Namespace, name, serviceAccount.Name, "BuildRun", buildRun.Name)
+		// add the secrets references into the new sa
+		ApplyCredentials(ctx, build, serviceAccount)
```

Comment:
`ApplyCredentials` does not do a save. You need to move this to before the `client.Create` call.

Reply:
good catch! thanks, fixed

Reply:
this deserves new integration tests., Im surprised this didnt break anything

Reply:
> Im surprised this didnt break anything

Me too. :-) I think integration and e2e test use an anonymous registry. Is not too difficult to change and we should do that probably soon.

Reply:
I extended the `creates a new service-account and deletes it after the build is terminated` to check if an output secret exists on the autogenerated sa